null.go

```diff
@@ -238,3 +238,70 @@ func (m *Map) UnmarshalJSON(data []byte) error {
 	}
 	return json.Unmarshal(data, &m.m)
 }
+
+// JSON is a json.RawMessage that will marshall as null when empty or nil.
+// null and {} values when unmashalled or scanned from a DB will result in a nil value
+type JSON json.RawMessage
+
+// Scan implements the Scanner interface for decoding from a database
+func (j *JSON) Scan(src interface{}) error {
+	if src == nil {
+		*j = nil
+		return nil
+	}
+
+	var source []byte
+	switch src.(type) {
+	case string:
+		source = []byte(src.(string))
+	case []byte:
+		source = src.([]byte)
+	default:
+		return fmt.Errorf("incompatible type for JSON type")
+	}
+
+	// 0 length string is same as nil
+	if len(source) == 0 || string(source) == "{}" {
```

Comment:
Why the need to treat `{}` the same as `null`? Not sure that makes sense on a general level. What about `[]` ?

Reply:
I think it is the same argument as saying `""` shouldn't exist in your DB. If you want empty, then store `NULL`. Agree that `[]` could be the same but that kinda felt like a bridge too far but maybe that's true as well. (`{}` feels more detaulty to me than `[]`)

It is kinda the impetus for this existing at all really, in that for example the FlowStart case, you are storing `parent_summary` and by default the sqlx JSON type will store NULL values as `{}` in the DB. Which then make switching on that case more complicated in the code, in that you have to check whether the `json.RawMessage` is non-empty AND not `{}` etc.. if you want to switch on whether that start has a parent or not. Having only one empty representation reduces potential for errors on the code level quite a bit.

Again, same argument as not storing empty strings in your DB.

Reply:
And I suppose then the argument for treating ONLY `{}` as NULL is that other things seem to write NULL as `{}` so that's the special case you want to deal with. In the end we make the rules so we can do whatever we want, but seemed like the right balance to me.

Reply:
sqlx's `types.JSONText` saving `json.RawMessage(nil)` as `{}` seems a little weird but I guess they had to pick a default that was actually valid JSON 🤷‍♂ 

A way this could trip us up is if someone was passing an extra dict to start flows `{"name": "Bob"}`, and they are using something like Go's `omitempty` so every now again when name isn't set, they end up passing `{}`. As long we preserve the fact that's an object, then in Excellent `trigger.params` is `{}` and `trigger.params["name"]` works fine even if name doesn't exist. If however we convert `trigger.params` to `nil` then that expression blows up.

It's not quite equivalent to ignorning the distinction between `NULL` and empty strings because there's only one possible non-NULL string value it makes sense to interpret `NULL` as... `""`

In a way, it would be better if `FlowStart.parent_summary` was non-NULL (and all of our JSON fields), i.e. it could be `null` or `{}` or `[]`, but it's always valid JSON. Never have worry about the `nil` case. Is that an option?

Reply:
Huh, I would suggest that trigger.params acting differently for `{}` and NULL for a case like `trigger.params["name"]` is wrong. IE, I would expect those to behave the same, that seems a bit too subtle for an end user to figure out no?

I disagree that having a non-null parent_summary makes sense. Again, then you get yourself in a state where you NEED to query both `{}` and `[]` instead of just `IS NULL`. Summary is even more clearly a NULL case to me because we control that field, so `[]` is never going to appear anyways and `{}` isn't a valid case.

Also in the go side of things then you are left doing things like:

```golang
if (start.ParentSummary() == "{}"){
}
```

vs the more natural:
```golang
if (start.ParentSummary() == nil){
}
```

Or perhaps most correct:
```golang
if (len(start.ParentSummary()) == 0){
}
```

Also note that the reason this comes up is that JSON representation matters here too and starts get Marshalled. So do we really want a `parent_summary` field on every field to be set to `{}` or do we want `parent_summary` to only exist in the case where it is set? To me the answer is clearly the latter.

Same with params on a start. If it isn't set then I expect that to be NULL in the DB and missing in the JSON.

Reply:
Actually that first case would have to be more like:

```golang
if (string(start.ParentSummary()) == "{}"){
}
```

Which is even more awkward.

Reply:
So maybe there's two different types of JSON field here. A field like `FlowStart.parent_summary` is something we control the schema of and we know it should only ever be in one of two states: `NULL` (i.e. not set), or a valid JSON object `{...}` (i.e. set). That instances of fields like that have ended up with `{}` values actually feels like a bug. It's not a meaningful state for a field like that. Were we supposed to be using `sqlx.NullJSONText`?

A field like `FlowStart.extra` gets its value from the user, is supposed to behave like arbitrary JSON, and it seems like `{}`, `[]`, `""` and `null` should all be fine there. If a user is passing in extra to flows I think we can assume they understand JSON types.

Trying to fix this by saying that for every JSON field, `{}` should always be read as `nil` seems.. like a hammer solution. And certainly diminishes the general usability of this library outside of mailroom.

Couldn't we switch to a new field type which doesn't serialize `nil` as `{}`? And then use a data migration to fix those fields which have invalid `{}` values? Then we can drop the annoying `== "{}"` checks in mailroom.

Reply:
To be clear, it isn't a giant hammer solution because not all JSON fields need to use this, it is obviously opt in.

That said, I think you've convinced me that treating `{}` specially is probably wrong, especially as the Python side of things doesn't do anything special there either. So I'll remove that treatment.

The trigger.params stuff is totally unrelated to how we marshal JSON to and fro the DB so opened another issue for that here: https://github.com/nyaruka/goflow/issues/738

Reply:
True not every field has to use this - may be it's worth having specialized JSON field types - like something that is specifically for object/`{..}` fields. Then special treatment for `{}` doesn't feel weird.